Add Vision page (#36)
NUbots/NUbook: src/book/02-system/01-nubots/05-vision.mdx

The parameters are used in the **Camera module** to find and set up the cameras. The Camera module emits [Image messages](https://github.com/NUbots/NUbots/tree/master/shared/message/input/Image.proto).

The [projection tool](https://github.com/NUbots/NUbots/tree/master/shared/utility/vision/projection.h), based on [panotools' fisheye projection calculations](https://wiki.panotools.org/Fisheye_Projection), maps a portion of the surface of a sphere to a flat image. The type of projection is specified by the above parameter `len.projection`.

# Visual Mesh

## Algorithm

The [Visual Mesh](https://doi.org/10.1007/978-3-030-27544-0_4) underpins the vision system and is used for sparse **detection** of balls, points on the field, field lines, goal posts and other robots. It is an **input transformation** that uses knowledge of a cameras orientation and position relative to an observation plane to increase the performance and accuracy of a **convolutional neural network**. It utilises the **geometry of objects** to create a **mesh structure** that gives high accuracy of detection regardless of the distance to the object.

> The **height** and **orientation** of the camera is tracked using the kinematics and inertial measurement unit of the robot. The **lens type** and the expected **radius** of any soccer ball are given as a configuration value. These values are used to create a set of **unit vectors** with the origin at the camera position. These unit vectors represent rays of light travelling towards the camera. Each vector is **mapped** to a **point on the image** using an appropriate lens projection equation. Two equations are to efficiently sample the space around the camera to obtain the **array of vectors** and associated **sample points**.

Reviewer: ```suggestion
The **height** and **orientation** of the camera is tracked using the kinematics and inertial measurement unit of the robot. The **lens type** and the expected **radius** of any soccer ball are given as a configuration value. These values are used to create a set of **unit vectors** with the origin at the camera position. These unit vectors represent rays of light travelling towards the camera. Each vector is **mapped** to a **point on the image** using an appropriate lens projection equation. Two equations are used to efficiently sample the space around the camera to obtain the **array of vectors** and associated **sample points**.
```

  $$
  r_a = \frac{r_b}{\sqrt{1 - r^2}}
  $$

  $$
  r_p = \lVert \text{axis} \cdot \frac{r_b - \text{rWCc.z}}{\text{axis.z}} + \text{rWCc} \rVert
  $$

  - $r_a$ is the angular-based size
  - $r_p$ is the projection-based size
  - $r_b$ is the actual ball radius we are expecting
  - $r$ is the calculated cluster radius
  - axis is the cone axis as specified above
  - rWCc is the vector from camera to world in camera space. For more information on this convention, see the [Mathematics](/system/nubots/mathematics#homogeneous-transformations) page.

- The cluster must have a distance less than the length of the field, as given by the FieldDescription message.

Any remaining clusters are assumed to be balls and so are emitted together as a [Balls message](https://github.com/NUbots/NUbots/blob/master/shared/message/vision/Ball.proto).

## Goal Detector

The [goal detector](https://github.com/NUbots/NUbots/blob/master/module/vision/GoalDetector/src/GoalDetector.cpp) follows a similar structure to the ball detector. It receives a GreenHorizon and a FieldDescription message.

> All goal post edge points (points that are goal points that have at least one neighbour that is **not** a goal post) are found and partitioned into clusters. Clusters are **discarded** if they fail to meet any of the following criteria

Reviewer: ```suggestion
All goal post edge points (points that are goal points that have at least one neighbour that is **not** a goal point) are found and partitioned into clusters. Clusters are **discarded** if they fail to meet any of the following criteria
```

description: Introduction to vision and how it works in the NUbots codebase.
slug: /system/nubots/vision
---

The vision system aims to convert images from a camera in the robot into useful information, such as ball position, goal position, field lines and obstacles. This information is used to determine behaviour and localisation. The vision system encompasses the [hardware](/system/hardware/overview/#cameras), object detection, and post processing.

# Cameras

The cameras have the following parameters that are used by the object detection algorithm.

| Parameter           | Type                 | Description                                                                                                                                       |
| ------------------- | -------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------- |

> | `serial_number`     | string               | The serial number of the camera.                                                                                                                  |

Reviewer: ```suggestion
| `serial_number`     | string               | The serial number of the camera. This is used to identify the camera and distinguish it from other cameras in the robot.                                                                                                                 |
```

> | `lens.projection`   | string               | The fisheye lens projection type. Can be rectilinear, equidistant or equisold.                                                                    |

Reviewer: ```suggestion
| `lens.projection`   | string               | The lens projection type. Can be rectilinear, equidistant or equisold.                                                                    |
```

These parameters are set for each camera as configuration values in the [Camera module](https://github.com/NUbots/NUbots/tree/master/module/input/Camera/data/config), in each robot's respective folder. The values for the left camera on the robot will be stored in `Left.yaml`. The values for the right camera on the robot will be stored in `Right.yaml`.

The parameters are used in the Camera module to find and set up the cameras. The Camera module emits [Image messages](https://github.com/NUbots/NUbots/tree/master/shared/message/input/Image.proto).

The [projection tool](https://github.com/NUbots/NUbots/tree/master/shared/utility/vision/projection.h), based on [panotools' fisheye projection calculations](https://wiki.panotools.org/Fisheye_Projection), maps a portion of the surface of a sphere to a flat image. The type of projection is specified by the above parameter `len.projection`.

# Visual Mesh

## Algorithm

The [Visual Mesh](https://doi.org/10.1007/978-3-030-27544-0_4) underpins the vision system, and is used for sparse detection of balls, points on the field, field lines, goal posts and other robots.

> The Visual Mesh is an input transformation that uses knowledge of a cameras orientation and position relative to an observation plane to greatly increase the performance and accuracy of a convolutional neural network. It utilises the geometry of objects to create a mesh structure that ensures that a similar number of samples points are selected regardless of distance. The result is that networks can be much smaller and simpler while still achieving high levels of accuracy. The Visual Mesh is very capable when it comes to detecting distant objects. Distant objects often become too small for a network to detect accurately, but as the Visual Mesh normalises, its size detections are still accurate. The Visual Mesh can detect soccer balls, field lines and goal posts.

Reviewer: ```suggestion
The Visual Mesh is an input transformation that uses knowledge of a cameras orientation and position relative to an observation plane to greatly increase the performance and accuracy of a convolutional neural network. It utilises the geometry of objects to create a mesh structure that ensures that a similar number of sample points are selected regardless of distance. The result is that networks can be much smaller and simpler while still achieving high levels of accuracy. The Visual Mesh is very capable when it comes to detecting distant objects. Distant objects often become too small for a network to detect accurately, but as the Visual Mesh normalises over distance, its size detections are still accurate. The Visual Mesh can currently detect soccer balls, field lines and goal posts.
```

The Visual Mesh is created using the following geometric assumptions for each image:

- The camera lens type, height and orientation with respect to the ground are known.
- The object is spherical with known radius and remains on the ground.

> The height, orientation and radius are used to create a set of unit vectors based at the camera position. Any vector with its origin at the camera can be thought of as a ray of light travelling towards the camera. If the vector is within the field of view of the camera it will be collected by the lens. The vector is mapped to a point on the image using the lens projection equation, provided the point is within the bounds of the image sensor of the camera. The Visual Mesh is formed by taking an array of vectors and associating each of them with a point and pixel in the image. The array of vectors is specially constructed using two equations to efficiently sample the space around the camera for the object. This means the sample points on the image will efficiently sample the image for the object.

Reviewer: ```suggestion
The height, orientation and radius are used to create a set of unit vectors based at the camera position. Any vector with its origin at the camera can be thought of as a ray of light travelling towards the camera. If the vector is within the field of view of the camera it will be collected by the lens. Vectors collected by the camera are mapped to points on the image using the lens projection equation. The Visual Mesh is formed by taking an array of vectors and associating each of them with a, possibly interpolated, pixel in the image. The array of vectors is specially constructed using two equations to efficiently sample the space around the camera for the object. This means the sample points on the image will efficiently sample the image for the object.
```

## Green Horizon Detector

When the [green horizon detector](https://github.com/NUbots/NUbots/blob/master/module/vision/GreenHorizonDetector/) receives a VisualMesh message, it uses the information in this message to create one large cluster of points that is determined to be the field.

The points are first filtered to give only potential field points. The first field point is added to a cluster with all its field point neighbours. Neighbouring field points are added to the cluster until all field points who are neighbours of the points in the cluster are added to the cluster.

Clustering is repeated until all field points are part of a cluster. Clusters that are smaller than a given threshold are discarded. Clusters are merged unless they overlap. If the clusters overlap, we keep the larger cluster and discard the smaller cluster.

After clustering has finished, one cluster will remain which should represent the field. An upper convex hull algorithm is applied to the final cluster to determine the edge of the field. The detector will emit a GreenHorizon message that specifies the location of the edge of the field.

## Ball Detector

> The [ball detector receives](https://github.com/NUbots/NUbots/blob/master/module/vision/BallDetector/) a GreenHorizon message and a [FieldDescription message](https://github.com/NUbots/NUbots/blob/master/shared/message/support/FieldDescription.proto). The FieldDescription message is output by the [SoccerConfig module](https://github.com/NUbots/NUbots/tree/master/module/support/configuration/SoccerConfig) which creates the FieldDescription message from configuration values specifying the layout of the soccer field.

Reviewer: ```suggestion
The [ball detector](https://github.com/NUbots/NUbots/blob/master/module/vision/BallDetector/) receives a GreenHorizon message and a [FieldDescription message](https://github.com/NUbots/NUbots/blob/master/shared/message/support/FieldDescription.proto). The FieldDescription message is output by the [SoccerConfig module](https://github.com/NUbots/NUbots/tree/master/module/support/configuration/SoccerConfig) which creates the FieldDescription message from configuration values specifying the layout of the soccer field as well as the radius of the ball being used.
```

The ball detector forms clusters out of all the points that the Visual Mesh has identified as ball points. Specifically, the clusters are formed from Visual Mesh points that are identified as being a ball point, but have at least $1$ neighbour which is **not** a ball point. This allows us to form clusters of ball edge points.

> Any clusters which are above the field edge (i.e. the green horizon) or are smaller than a given threshold are discarded. A circular cone is then fitted to each cluster. The cone axis is determined from the line segment between the centre on the camera and the average of all ball edge points. The radius of the cone is determined by the maximum distance between the ball edge points and the average of all of the ball edge points.

Reviewer: ```suggestion
Any clusters which are above the field edge (i.e. the green horizon) or are smaller than a given threshold are discarded. A circular cone is then fitted to each cluster. The cone axis is determined from the line segment between the centre on the camera and the centre of the ball, determined by averaging all of the ball edge points. The radius of the cone is determined by the maximum distance between the ball edge points and the centre of the ball.
```

> Degree of circle fit is used to determine if the cluster fits the shape of a circle well enough to be a ball. If it does not, then we discard it. Any ball clusters that are too close, as defined by some configuration variable, are also discarded. Clusters with a large difference between their angular and projection based distances are discarded. Any balls further away than the length of the field, as specified by the FieldDescription, are discarded.

Reviewer: Might be worth turning this paragraph into a list and maybe adding some of the equations as well?

Reviewer 2: Also, what's this talk about fitting the shape of a circle? We had a whole thing about this!

Reviewer: The base of a circular cone is still a circle, even in 3D

Any clusters that are still remaining are determined to be balls and so are emitted as a [Balls message](https://github.com/NUbots/NUbots/blob/master/shared/message/vision/Ball.proto).

## Goal Detector

The [goal detector](https://github.com/NUbots/NUbots/blob/master/module/vision/GoalDetector/src/GoalDetector.cpp) follows a similar structure to the ball detector. It receives a GreenHorizon and a FieldDescription message.

> Any points that are not goal points or are completely surrounded by goal points are discarded. Any points above the green horizon are also discarded. The points we have left are goal post edge points. These points are clustered. Any clusters that do not intersect the green horizon are discarded, since we assume goals are higher than the field. Any clusters that are smaller than a given threshold are also discarded.

Reviewer: ```suggestion
All goal post edge points (point that are goal points that have at least one neighbour that is **not** a goal post) are found and partitioned into clusters. Any clusters that do not intersect the green horizon are discarded, since we assume goals are higher than the field. Any clusters that are smaller than a given threshold are also discarded.
```
